Histogram (revisited)

LukeMathWalker opened the pull request:
Based on our discussion in #8, I have revised the implementation.

It needs a testing suite and a couple more helper methods on `HistogramCounts` but the skeleton it's there. Let me know your thoughts @jturner314

jturner314 commented:
I think this is a good approach.

A few thoughts:

* `Edges` needs to be careful about repeated elements. There are two possible strategies:

  1. `Edges` can make sure all of the elements are unique, in addition to being in order. The constructor could do this by removing duplicates or by returning an error when there are duplicates. This would make the rest of the implementation simpler but would be less convenient for the user.

  2. Alternatively, it needs to be careful about identifying the correct bin. For example, the current implementation of `.indexes()` doesn't correctly handle repeated edges, since in the case of repeated elements, `.binary_search()` can return the index of any of those elements.

     Assuming this strategy isn't much more difficult to implement, I prefer it (to make things simple for the user). It's also worth noting that NumPy uses this strategy.

* I'd recommend removing the `ndim` member from `HistogramCounts` and instead adding a `.ndim()` method that calls `self.counts.ndim()`. (This means there are fewer things to keep in-sync.)

* In the future, it would be worth adding the dimension of `counts` to the type of `HistogramCounts` instead of always using `ArrayD`. This isn't necessary for a first implementation, though.

* In #8, I suggested separate `HistogramCounts` and `HistogramDensity` types. On further reflection, I don't think a `HistogramDensity` type is worth adding. Instead, I'd suggest adding a `.density()` method on `HistogramCounts` that returns an array of densities. This does mean that if all you want is density, there's a conversion cost from counts -> densities once all the data have been counted, but I think the simplicity is worth the cost (and this approach may in-fact be cheaper anyway). (This also means that we can rename `HistogramCounts` to just `Histogram`.)

LukeMathWalker commented:
I agree on almost all points - I think it makes sense for `Edges` to drop duplicates in the constructor (without panicking or raising an error): we pay a little more upfront (`O(n)` where `n` is the number of bins) but we benefit from faster `add_observation` calls, which is where the bulk of the computation will be happening. 

I'll work on the other points and then I think we can merge.

jturner314 commented:
> we benefit from faster `add_observation` calls

I'm not sure the difference would be significant. I think all we'd need to do is change

```rust
Ok(i) => Some((i, i + 1)),
```

to

```rust
Ok(mut i) => {
    while i + 1 < n_edges && self.edges[i] == self.edges[i + 1] {
        i += 1;
    }
    Some((i, i + 1))
}
```

which, if all the elements are unique, requires only two extra comparisons in this branch (and this branch is executed only when the value equals an edge).

That said, I'm fine with requiring unique edges for the time being. We can benchmark later to see if it really matters.

LukeMathWalker commented:
I agree that the performance gain is probably minimal, given that it is only triggered when we are adding an edge as an observation, but I don't see any real benefit from a user perspective in keeping duplicated edges around.

I was trying to write a test on `add_observation` to assert that the `Histogram::counts` matrix was equal to `array![[0, 1], [0, 0]]` but the compiler of course complained that one is an instance of `ArrayD` and the other one is an instance of `Array2`.
Is there any way to convert a fixed-dimension array to a dynamically sized array in `ndarray`? I tried to have a look in the docs, but I couldn't find anything related to it.

With respect to this PR:
- What strategy would you suggest to add a dimension parameter to `Histogram`?
- Would it make sense to introduce a `Grid<A>` struct to encapsulate `Vec<Bins<A>>` and provide a bunch of helper methods?

jturner314 commented:
> I don't see any real benefit from a user perspective in keeping duplicated edges around.

My primary concern is a user writing code that assumes that the number of edges in the histogram is the same as the number of edges they provided.

I agree that deduplicating edges in the `Edges` constructor is reasonable behavior, though. We just need to make it clear in the docs that this is what we're doing so that the user doesn't assume the number of edges is unchanged.

> Is there any way to convert a fixed-dimension array to a dynamically sized array in `ndarray`?

Is [`.into_dyn()`](https://docs.rs/ndarray/0.12.0/ndarray/struct.ArrayBase.html#method.into_dyn) what you're looking for?

> What strategy would you suggest to add a dimension parameter to `Histogram`?

I'm thinking something along the lines of this:

```rust
pub struct Histogram<A: Ord, D: Dimension> {
    counts: Array<usize, D>,
    bins: Vec<Bins<A>>,
}

impl<A: Ord, D: Dimension> Histogram<A, D> {
    /// **Panics** if the length of `bins` does not match `D`.
    pub fn new(bins: Vec<Bins<A>>) -> Self {
        let mut shape = D::zero_index_with_ndim(bins.len());
        bins.iter().zip(shape.slice_mut()).for_each(|b, s| *s = b.len());
        let counts = Array::zeros(shape);
        Histogram { counts, bins }
    }
}

pub trait HistogramExt<A, S>
where
    S: Data<Elem = A>,
{
    /// **Panics** if `d` is different from `bins.len()` or if it does not match `Dh`.
    fn histogram<Dh>(&self, bins: Vec<Bins<A>>) -> Histogram<A, Dh>
    where
        A: Ord;
}
```

The implementation for `new` uses some undocumented methods of the `Dimension` trait. I don't see any good way to do it otherwise; I can modify `ndarray` to expose these.

Usage looks like this:

```rust
// dynamic dimension
let hist = arr.histogram::<IxDyn>(bins);
// or, if we define type aliases,
let hist: HistogramD<_> = arr.histogram(bins);
// fixed dimension; panics if `bins.len() != 2` or if `arr.len_of(Axis(1)) != 2`
let hist = arr.histogram::<Ix2>(bins);
```

> Would it make sense to introduce a `Grid<A>` struct to encapsulate `Vec<Bins<A>>` and provide a bunch of helper methods?

I think so. For example, as a user, it would be nice to be able to easily figure out the bin corresponding to a data point (using a method on `Grid`), and then look up the corresponding count. We could just place all these methods on `Histogram` instead of a separate `Grid` type, but I could see there being cases where the user would want to decompose the `Histogram` into separate `Grid` and `Array` pieces.

LukeMathWalker commented:
I have added strategies to infer the optimal bin edges as well as a builder to get a grid using those strategies on each coordinate axis (right now it is constrained to use the same strategy across all axes, but we can make it more versatile in the future).

I have documented all methods and added key docs example where needed. 

I think we are good to go to merge it as a first implementation @jturner314.

LukeMathWalker commented on src/histogram/grid.rs:
```diff
@@ -0,0 +1,180 @@
+use std::marker::PhantomData;
+use ndarray::{ArrayView1, ArrayView2, Axis};
+
+/// A `Grid` is a partition of a rectangular region of an `n`-dimensional
```
I have to learn this Markdown tricks! Much much better!

jturner314 replied:
I used a combination of HTML (the `<sub>subscript</sub>` for subscripts), Unicode characters (the multiplication sign, ellipsis, and em dashes), and `*italics*` for italics.

LukeMathWalker commented on src/quantile.rs:
```diff
@@ -410,3 +410,83 @@ where
+        A: Ord + Clone,
+        S: DataMut,
+        I: Interpolate<A>,
+    {
```
It's a quite convoluted notation - I think it makes sense to provide a specialized method for 1d array. I was actually considering to rewrite the `quantile_axis_mut` implementation to use the 1d implementation on each 1d lane.

jturner314 replied:
Sorry, I wasn't clear. I'm suggesting to replace the implementation of `quantile_mut` with `self.quantile_axis_mut::<I>(Axis(0), q)[()].clone()` to reduce duplication, not remove the method.

jturner314 replied:
You could certainly go the other way, too (instead of rewriting `quantile_mut` in terms of `quantile_axis_mut`, rewrite `quantile_axis_mut` in terms of `quantile_mut`, as you suggested).

LukeMathWalker replied:
For now I'll use `quantile_axis_mut` to implement `quantile_mut` and I'll review which way is more convenient when I get to implement optimized computations of multiple quantiles.

LukeMathWalker commented on src/histogram/histograms.rs:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    /// Borrows a view on the histogram matrix.
+    pub fn as_view(&self) -> ArrayViewD<usize> {
```
It makes sense, it leaves us semantic room to add `density`.

LukeMathWalker commented on src/histogram/bins.rs:
````diff
@@ -0,0 +1,435 @@
+    ///     vec![0, 3, 5].as_slice()
+    /// );
+    /// ```
+    pub fn as_slice(&self) -> &[A] {
````
I agree, but how do I return a view to an array that I am not storing? I can't get around the compiler complaining that the value does not live long enough.

jturner314 replied:
You can do this:

```rust
    pub fn as_array_view(&self) -> ArrayView1<A> {
        ArrayView1::from(&self.edges)
    }
```

With explicit lifetimes, this is equivalent to

```rust
    pub fn as_array_view<'a>(&'a self) -> ArrayView1<'a, A> {
        ArrayView1::from(&self.edges)
    }
```

This uses the [`From` implementation for `ArrayView1`](https://docs.rs/ndarray/0.12.0/ndarray/type.ArrayView.html#method.from), which is what we want in this case. A related method is [`from_shape`](https://docs.rs/ndarray/0.12.0/ndarray/type.ArrayView.html#method.from_shape), which works for n-D views.

LukeMathWalker commented on src/histogram/grid.rs:
```diff
@@ -0,0 +1,180 @@
+///
+/// // The optimal grid layout is inferred from the data,
+/// // specifying a strategy (Auto in this case)
+/// let grid = GridBuilder::<usize, Auto<usize>>::from_array(observations.view()).build();
```
There was a bug in Freedman Diaconis - I fixed it.

LukeMathWalker commented:
I managed to address all comments (and I fixed the bug you spotted) - they were all on point, thanks for taking the time to go through the PR with that level of attention!

I have also reduced the number of types parameter on `BinsBuildingStrategy` from 2 to 1 using the strategy you suggested (associated type on trait).

With respect to the issue of bin boundaries I proceeded as follows: all strategies now provide an optimal bin width (either directly through the rule they specify or indirectly recasting the optimal number of bins to an optimal bin width) and we make sure that the last edge is strictly greater than the maximum of the array that has been passed to the builder. 
This basically means that we might add an extra bin to the right if it is required to account for the maximum value, but we don't modify the grid parameters to achieve it. What do you think?

jturner314 commented:
> I managed to address all comments (and I fixed the bug you spotted) - they were all on point, thanks for taking the time to go through the PR with that level of attention!

You're welcome. Thanks for working on this! Would you like me to review the updated version, or do you think it's ready to merge?

> This basically means that we might add an extra bin to the right if it is required to account for the maximum value, but we don't modify the grid parameters to achieve it. What do you think?

I think that's fine.

For future versions, I think it would be worth investigating how Julia's [StatsBase](https://juliastats.github.io/StatsBase.jl/latest/empirical.html#Histograms-1) does things, because StatsBase is similar to our implementation in using all half-open intervals (unlike NumPy, which considers the last bin to be a closed interval).

LukeMathWalker commented:
Let me add a couple of tests and then I think we are good to go!
On 13 Nov 2018, at 19:21, Jim Turner <notifications@github.com<mailto:notifications@github.com>> wrote:

I managed to address all comments (and I fixed the bug you spotted) - they were all on point, thanks for taking the time to go through the PR with that level of attention!

You're welcome. Thanks for working on this! Would you like me to review the updated version, or do you think it's ready to merge?

This basically means that we might add an extra bin to the right if it is required to account for the maximum value, but we don't modify the grid parameters to achieve it. What do you think?

I think that's fine.

For future versions, I think it would be worth investigating how Julia's StatsBase<https://juliastats.github.io/StatsBase.jl/latest/empirical.html#Histograms-1> does things, because StatsBase is similar to our implementation in using all half-open intervals (unlike NumPy, which considers the last bin to be a closed interval).

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/jturner314/ndarray-stats/pull/9#issuecomment-438402131>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ATyLWDO3Nh68h4dbfGUSCFMtE_czfAi7ks5uuxu3gaJpZM4Xs2FI>.

jturner314 commented on src/quantile.rs:
```diff
@@ -410,3 +410,57 @@ where
+        S: DataMut,
+        I: Interpolate<A>,
+    {
+        self.quantile_axis_mut::<I>(Axis(0), q)[()].clone()
```
Now that rust-ndarray/ndarray#535 is merged, we can change

```suggestion
        self.quantile_axis_mut::<I>(Axis(0), q).into_scalar()
```
